vhdx/src/vhdx_bat.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright © 2021 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+
+use crate::vhdx_header::RegionTableEntry;
+use crate::vhdx_metadata::DiskSpec;
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+use remain::sorted;
+use std::fmt::{self, Display};
+use std::fs::File;
+use std::io::{self, Seek, SeekFrom};
+use std::mem::size_of;
+
+// Payload BAT Entry States
+pub const PAYLOAD_BLOCK_NOT_PRESENT: u64 = 0;
+pub const PAYLOAD_BLOCK_UNDEFINED: u64 = 1;
+pub const PAYLOAD_BLOCK_ZERO: u64 = 2;
+pub const PAYLOAD_BLOCK_UNMAPPED: u64 = 3;
+pub const PAYLOAD_BLOCK_UNMAPPED_V095: u64 = 5;
+pub const PAYLOAD_BLOCK_FULLY_PRESENT: u64 = 6;
+pub const PAYLOAD_BLOCK_PARTIALLY_PRESENT: u64 = 7;
+
+// mask for the BAT state
+pub const BAT_STATE_BIT_MASK: u64 = 0x07;
+// mask for the offset within the file in units of 1 MB
+pub const BAT_FILE_OFF_MASK: u64 = 0xFFFFFFFFFFF00000;
+
+macro_rules! div_round_up {
+    ($n:expr,$d:expr) => {{
+        ($n + $d - 1) / $d
+    }};
+}
+
+#[sorted]
+pub enum Error {
+    // Invalid BAT entry
+    InvalidBatEntry,
+    // Invalid number of BAT entries
+    InvalidEntryCount,
+    // Failed to read BAT from file
+    ReadBat(io::Error),
+    // Failed to write BAT entries to file
+    WriteBat(io::Error),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl Display for Error {
+    #[remain::check]
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        #[sorted]
+        match self {
+            InvalidBatEntry => write!(f, "Invalid BAT entry"),
+            InvalidEntryCount => write!(f, "Invalid BAT entry count"),
+            ReadBat(e) => write!(f, "Failed to read BAT entry {}", e),
+            WriteBat(e) => write!(f, "Failed to write BAT entry {}", e),
+        }
+    }
+}
+
+#[derive(Default, Clone, Debug)]
+pub struct BatEntry {
+    pub bat_entry: u64,
```

Comment:
Why do you need this?

Reply:
This represents the BAT Entry structure described in the spec.

Reply:
You can use the following notation:
```rust
pub struct BatEntry(u64);
```
And then you would access the value from the code the following way:
```rust
let my_bat_entry = BatEntry(10);
println!("my_bat_entry value: {}", my_bat_entry.0)
```

Reply:
Why initializing with 10? Are you suggesting to replace the vector that contains bat entries as well?

Reply:
This was just an example of how to use the syntax :)

Reply:
Got it!